src/main/scala/com/microsoft/hyperspace/index/ExperimentalStructure.scala

```diff
@@ -0,0 +1,148 @@
+      files.toArray.groupBy(_.getPath.getParent)
+    val rootPath = getRoot(leafDirToChildrenFiles.head._1)
+
+    val subDirs = leafDirToChildrenFiles
+      .filterNot(_._1.isRoot)
+      .map {
+        case (dir, statuses) =>
+          createDirectoryTreeWithoutRoot(dir, statuses)
+      }
+      .toSeq
+
+    val rootFiles: Seq[FileInfo1] =
+      leafDirToChildrenFiles.getOrElse(rootPath, Array()).map(FileInfo1(_))
+    NewContent(Directory(rootPath.toString, rootFiles, subDirs))
+  }
+
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
+  }
+
+  private def createDirectoryTreeWithoutRoot(
+      dirPath: Path,
+      statuses: Array[FileStatus]): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = statuses.map(FileInfo1(_)))
+    } else {
+      createTopLevelNonRootDirectory(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = statuses.map(FileInfo1(_))))
+    }
+  }
+
+  private def createTopLevelNonRootDirectory(dirPath: Path, child: Directory): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = Seq(), subDirs = Seq(child))
+    } else {
+      createTopLevelNonRootDirectory(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = Seq(), subDirs = Seq(child)))
+    }
+  }
+}
+
+case class Directory(name: String, files: Seq[FileInfo1] = Seq(), subDirs: Seq[Directory] = Seq())
+
+object Directory {
+  def apply(path: Path): Directory = {
+    val files: Seq[FileInfo1] = {
+      try {
+        val fs = path.getFileSystem(new Configuration)
+        val statuses = fs.listStatus(path).filter(status => isDataPath(status.getPath))
+
+        // Assuming index directories don't contain nested directories. Only Leaf files.
+        assert(statuses.forall(!_.isDirectory))
+
+        statuses.map(FileInfo1(_)).toSeq
+      } catch {
+        // FileNotFoundException is an expected exception before index gets created.
+        case _: FileNotFoundException => Seq()
+        case e: Throwable => throw e
+      }
+    }
+    if (path.isRoot) {
+      Directory(path.toString, files, Seq())
+    } else {
+      Directory(path.getParent, Directory(path.getName, files, Seq()))
+    }
+  }
+
+  /* from PartitionAwareFileIndex */
+  private def isDataPath(path: Path): Boolean = {
+    val name = path.getName
+    !((name.startsWith("_") && !name.contains("=")) || name.startsWith("."))
+  }
+
+  def apply(path: Path, child: Directory): Directory = {
+    if (path.isRoot) {
+      Directory(path.toString, Seq(), Seq(child))
+    } else {
+      Directory(path.getParent, Directory(path.getName, Seq(), Seq(child)))
+    }
+  }
+}
+
+case class FileInfo1(name: String, size: Long, modifiedTime: Long)
+
+object FileInfo1 {
+  def apply(s: FileStatus): FileInfo1 =
+    FileInfo1(s.getPath.getName, s.getLen, s.getModificationTime)
+}
+
+object Main extends App {
```

Comment:
is this still WIP?

Reply:
Thanks @imback82 , I have cleaned up the code. I will continue to find more places to improve, but as of now it's ready.